libnetwork/resolver.go

```diff
@@ -448,25 +466,29 @@ func (r *Resolver) dialExtDNS(proto string, server extDNSEntry) (net.Conn, error
 	return extConn, nil
 }
 
-func (r *Resolver) forwardExtDNS(proto string, query *dns.Msg) *dns.Msg {
+func (r *Resolver) forwardExtDNS(ctx context.Context, proto string, query *dns.Msg) *dns.Msg {
+	ctx, span := otel.Tracer("").Start(ctx, "resolver.forwardExtDNS")
+	defer span.End()
+
 	for _, extDNS := range r.extDNSList {
 		if extDNS.IPStr == "" {
 			break
 		}
 
 		// limits the number of outstanding concurrent queries.
-		ctx, cancel := context.WithTimeout(context.Background(), extIOTimeout)
+		ctx, cancel := context.WithTimeout(ctx, extIOTimeout)
 		err := r.fwdSem.Acquire(ctx, 1)
 		cancel()
+
 		if err != nil {
 			r.logInverval.Do(func() {
-				r.log().Errorf("[resolver] more than %v concurrent queries", maxConcurrent)
+				r.log(ctx).Errorf("[resolver] more than %v concurrent queries", maxConcurrent)
 			})
```

Comment:
Now that there is a parent context, it's possible for the semaphore acquire to fail due to the parent context being done, not because of a timeout. I'm not too picky with how you want to handle other error values, but the "more than %v concurrent queries" log should be logged if and only if `errors.Is(err, context.DeadlineExceeded)`.